crypto/src/memo.rs

```diff
@@ -16,12 +23,107 @@ impl Default for MemoPlaintext {
 }
 
 impl MemoPlaintext {
-    pub fn encrypt(&self, _key: &Address) -> MemoCiphertext {
-        // TODO!
-        let memo_bytes = [0u8; MEMO_CIPHERTEXT_LEN_BYTES];
-        MemoCiphertext(memo_bytes)
+    /// Encrypt a memo, returning its ciphertext.
+    pub fn encrypt(
+        &self,
+        esk: &ka::Secret,
+        transmission_key: &ka::Public,
+        diversified_generator: &decaf377::Element,
+    ) -> Result<MemoCiphertext, anyhow::Error> {
+        let epk = esk.diversified_public(diversified_generator);
+        let shared_secret = esk
+            .key_agreement_with(&transmission_key)
+            .map_err(|_| anyhow!("could not perform key agreement"))?;
```

Comment:
Could this take an address type or something, so that we know the transmission key is valid, and we can treat the key agreement as infallible?